Code/Tools/AssetProcessor/native/AssetManager/PathDependencyManager.cpp

```diff
@@ -267,71 +359,39 @@ namespace AssetProcessor
                 }
 
                 // All checks passed, this is a valid dependency we need to save to the db
-                dependencyContainer.push_back();
-                auto& entry = dependencyContainer.back();
+                AzToolsFramework::AssetDatabase::ProductDependencyDatabaseEntry entry;
 
                 entry.m_productDependencyID = dependencyId;
                 entry.m_productPK = productDependencyDatabaseEntry.m_productPK;
                 entry.m_dependencySourceGuid = sourceEntry.m_sourceGuid;
                 entry.m_dependencySubID = matchedProduct.m_subID;
                 entry.m_platform = productDependencyDatabaseEntry.m_platform;
 
+                dependencyContainer.insert(entry);
```

Comment:
Did you test/consider to leave the vectors and deduplicate at the end? Considering the size of `ProductDependencyDatabaseEntry`, the hashing and comparison is likely expensive

Reply:
I'm not sure how that would help.  If I build a vector with duplicates in it and then copy it into a set, its still going to have to go through the same number of entries and hash them.  Or am I misunderstanding something?

Reply:
Well you can "move" the entry into the set to prevent a copy

Reply:
the move will save the copies of the two strings in it, but all the other members are simple types that dont move. Moving can provide an improvement there.

My concern was that the `ProductDependencyDatabaseEntry` type is not "easily" hashable and the comparison (which happens in a hash collision) is also expensive.
I am also not fully understanding how we can end up with duplicates here, one of the members is the dependency source guid. That leads me to think that for the same source, we end up adding the same dependency multiple times. Are we doing multiple passes?. I would expect that we traverse that DAG once

Reply:
This was due to sources/products with the same name, however I realized we're supposed to be filtering to just the source/product based on the dependency.  I've updated it to fix that issue and removed the set.